x/feegrant/key.go

```diff
@@ -22,30 +25,64 @@ const (
 
 var (
 	// FeeAllowanceKeyPrefix is the set of the kvstore for fee allowance data
+	// - 0x00<allowance_key_bytes>: allowance
 	FeeAllowanceKeyPrefix = []byte{0x00}
+
+	// FeeAllowanceQueueKeyPrefix is the set of the kvstore for fee allowance keys data
+	// - 0x01<allowance_prefix_queue_key_bytes>: <empty value>
+	FeeAllowanceQueueKeyPrefix = []byte{0x01}
 )
 
 // FeeAllowanceKey is the canonical key to store a grant from granter to grantee
 // We store by grantee first to allow searching by everyone who granted to you
+//
+// Key format:
+// - <0x00><len(grantee_address_bytes)><grantee_address_bytes><len(granter_address_bytes)><granter_address_bytes>
 func FeeAllowanceKey(granter sdk.AccAddress, grantee sdk.AccAddress) []byte {
+	fmt.Println("granter", address.MustLengthPrefix(granter.Bytes()))
+	fmt.Println("grantee", address.MustLengthPrefix(grantee.Bytes()))
+	fmt.Println("-----------------------------------------------------")
 	return append(FeeAllowancePrefixByGrantee(grantee), address.MustLengthPrefix(granter.Bytes())...)
 }
 
 // FeeAllowancePrefixByGrantee returns a prefix to scan for all grants to this given address.
+//
+// Key format:
+// - <0x00><len(grantee_address_bytes)><grantee_address_bytes>
 func FeeAllowancePrefixByGrantee(grantee sdk.AccAddress) []byte {
 	return append(FeeAllowanceKeyPrefix, address.MustLengthPrefix(grantee.Bytes())...)
 }
 
+// FeeAllowancePrefixQueue is the canonical key to store grant key.
+//
+// Key format:
+// - <0x01><exp_bytes><len(grantee_address_bytes)><grantee_address_bytes><len(granter_address_bytes)><granter_address_bytes>
+func FeeAllowancePrefixQueue(exp *time.Time, key []byte) []byte {
+	allowanceByExpTimeKey := AllowanceByExpTimeKey(exp)
+	return append(allowanceByExpTimeKey, key...)
+}
+
+// AllowanceByExpTimeKey returns a key with `FeeAllowanceQueueKeyPrefix`, expiry
+//
+// Key format:
+// - <0x01><exp_bytes>
+func AllowanceByExpTimeKey(exp *time.Time) []byte {
+	// no need of appending len(exp_bytes) here, `FormatTimeBytes` gives const length everytime.
+	return append(FeeAllowanceQueueKeyPrefix, sdk.FormatTimeBytes(*exp)...)
+}
+
+// ParseAddressesFromFeeAllowanceKey exrtacts and returns the granter, grantee from the given key.
+// Note: do not send the key with store prefix, remove the store prefix (first byte) while sending.
```

Comment:
It's a bit weird that this is the only function in this file that doesn't need the 1st byte

Reply:
we are iterating prefixStore in `AllowancesByGranter` that doesn't include the prefix in key.